src/main/scala/amba/apb/ToTL.scala

```diff
@@ -45,23 +45,41 @@ class APBToTL()(implicit p: Parameters) extends LazyModule
       require (edgeOut.manager.minLatency >= 1)
       assert (!(in.psel && !in.penable && out.d.valid))
 
+
+      val beat = TransferSizes(1, beatBytes)
+      // The double negative here is to work around Chisel's broken implementation of widening ~x.
+      val alignedAddr =  ~(~in.paddr | (beatBytes-1).U)
+      val dataSize = UInt(log2Ceil(in.params.dataBits/8))
+
+      // Is this access allowed? Illegal addresses are a violation of tile link protocol.
+      // If an illegal address is provided, return an error instead of sending over tile link.
+      val a_legal =
+        Mux(in.pwrite,
+          edgeOut.manager.supportsPutPartialSafe(alignedAddr, dataSize, Some(beat)),
+          edgeOut.manager.supportsGetSafe       (alignedAddr, dataSize, Some(beat)))
+
       val in_flight_reg = RegInit(false.B)
-      out.a.valid := in.psel && !in_flight_reg
-      in.pready := out.d.valid
+      val error_in_flight_reg = RegInit(false.B)
+      val in_flight = in_flight_reg || error_in_flight_reg
+
+      out.a.valid := in.psel && !in_flight && a_legal
+      in.pready := out.d.valid  || error_in_flight_reg
       out.d.ready := true.B
 
       when (out.a.fire()){in_flight_reg := true.B}
       when (out.d.fire()){in_flight_reg := false.B}
 
+      when (in.psel && !error_in_flight_reg){error_in_flight_reg := !a_legal}
+      when (error_in_flight_reg){ error_in_flight_reg := false.B }
+
```

Comment:
I agree that your formulation is correct. However, even though the spec guarantees it, I don't think we should rely on masters to hold paddr constant if we don't have to.

error_in_flight_reg := false.B
when (in.psel && !in_flight) {error_in_flight_reg := !a_legal}